[Cocoa] Report network transferred length instead of decoded buffer size.

basuke opened the pull request:
#### 1b949aac3e4f06f538dfbca9b93c2bc7db68742a
<pre>
[Cocoa] Report network transferred length instead of decoded buffer size.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287591">https://bugs.webkit.org/show_bug.cgi?id=287591</a>
<a href="https://rdar.apple.com/144530889">rdar://144530889</a>

Reviewed by Ben Nham.

ResourceMonitor wants to use precise network usage for accuracy purposes. Add implementation
to track the network transferred bytes in NetworkDataTask.

* LayoutTests/http/tests/iframe-monitor/compressed-resource-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/compressed-resource.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/compressed.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/generate-byte.py:
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::didReceiveData):

Canonical link: <a href="https://commits.webkit.org/290353@main">https://commits.webkit.org/290353@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d5a34477940d08c20ef56a91ef4ceb0726c53097

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89759 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/9288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/44645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/94752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/40527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91811 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/9675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17565 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/94752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/40527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92760 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/9675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/44645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/94752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/9675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/44645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39661 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/9675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/44645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96579 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16941 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12444 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17197 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/44645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/77324 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19082 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21762 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/44645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10120 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16954 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16695 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20147 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18477 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -387,6 +387,15 @@ static float toNSURLSessionTaskPriority(WebCore::ResourceLoadPriority priority)
 void NetworkDataTaskCocoa::didCompleteWithError(const WebCore::ResourceError& error, const WebCore::NetworkLoadMetrics& networkLoadMetrics)
 {
+    if (networkLoadMetrics.responseBodyBytesReceived > totalBytesTransferredOverNetwork()) {
+        WTFEmitSignpost(m_task.get(), DataTask, "received 0 bytes");
```
This doesn't seem to be the right string to associate with this signpost interval.

I think I'd leave out the `WTFEmitSignpost` call entirely here.

basuke replied:
But it may call `didReceiveData()` so treating it in part of completed sounds wrong.

basuke commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -387,6 +387,15 @@ static float toNSURLSessionTaskPriority(WebCore::ResourceLoadPriority priority)
+        if (m_client)
+            m_client->didReceiveData(WebCore::SharedBuffer::create());
+    }
+
```
Adding this workaround for older network stack, the test was happy with Ventura machines. But this workaround postpone the network usage metrics monitoring of ResourceMonitor until the end of the communication. If the resource is much bigger than threshold, unloading won't happen until the completion.

bnham replied:
Hopefully this extra IPC does not cause any issues...

Presumably this is because we folded the `bytesTransferredOverNetwork` (which is actually a delta) argument in to the `DidReceiveData` IPC.

I think a cleaner refactoring might be to move all of the network resource delta computation logic out of NetworkProcess and in to WebProcess, e.g. by:

- Changing `WebResourceLoader::DidReceiveData` IPC to report the total bytes transferred over the network over this load so far rather than a delta.
- When processing the `WebResourceLoader::DidFinishResourceLoad` (which already gets a `NetworkLoadMetrics` object), do any final delta computations there and report that final delta to the resource monitor.

This would get rid of the extra IPC and is probably easier to reason about. Probably does not need to block this patch.

basuke replied:
No that doesn't work because we need to distinguish cached resource, blobs and data urls which won't consume network usage. We need to track this in Network Process.

basuke replied:
Ah, sorry, you are saying using `NetworkLoadMetrics`. But that prevents measuring during the communication which is new stack is supporting.

basuke replied:
Okay, I agree with your idea. I'll do that in the following patch. That way is better way for other platforms.

basuke committed:
[Cocoa] Report network transferred length instead of decoded buffer size.
https://bugs.webkit.org/show_bug.cgi?id=287591
rdar://144530889

Reviewed by Ben Nham.

ResourceMonitor wants to use precise network usage for accuracy purposes. Add implementation
to track the network transferred bytes in NetworkDataTask.

* LayoutTests/http/tests/iframe-monitor/compressed-resource-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/compressed-resource.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/compressed.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/generate-byte.py:
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::didReceiveData):

Canonical link: https://commits.webkit.org/290353@main
